0_portfolio_input_check_functions.R

```diff
@@ -1092,6 +1092,142 @@ process_raw_portfolio <- function(portfolio_raw,
+  } else {
+    portfolio_total <- as_tibble(portfolio)
+    portfolio_total$direct_holding <- TRUE
+  }
+  
+  
+  fund_portfolio_total <- portfolio_total %>% filter(!is.na(fund_isin))
+  
+  fund_portfolio_total_mapped_value_usd <- fund_portfolio_total  %>% 
+    group_by(holding_id) %>% 
+    summarize(total_mapped_value_usd = sum(value_usd))
+  
+  
+  fund_portfolio_missing_value_usd <- fund_portfolio_total  %>% 
+    filter(nchar(isin)!=12) %>% 
+    group_by(holding_id) %>% 
+    summarize(missing_value_usd = sum(value_usd))
+  
+  
+  fund_portfolio_funds_in_funds_not_mapped_value_usd <-  fund_portfolio_total  %>% 
+    filter(nchar(isin)==12 & asset_type == "Funds") %>% 
+    group_by(holding_id) %>% 
+    summarize(funds_in_funds_not_mapped = sum(value_usd))
+  
+  
+  
+  fund_portfolio <- fund_portfolio_raw
+  fund_portfolio <- left_join(fund_portfolio, fund_portfolio_total_mapped_value_usd)
+  fund_portfolio$total_mapped_value_usd[is.na(fund_portfolio$total_mapped_value_usd)] <- 0
+  
+  
+  fund_portfolio <- left_join(fund_portfolio, fund_portfolio_missing_value_usd)
+  fund_portfolio$missing_value_usd[is.na(fund_portfolio$missing_value_usd)] <- 0
+  
+  fund_portfolio <- left_join(fund_portfolio, fund_portfolio_funds_in_funds_not_mapped_value_usd)
+  fund_portfolio$funds_in_funds_not_mapped[is.na(fund_portfolio$funds_in_funds_not_mapped)] <- 0
+  
+  
+  
+  fund_portfolio <- fund_portfolio %>% mutate(effective_coverage = (total_mapped_value_usd - missing_value_usd - funds_in_funds_not_mapped)/value_usd)
+  fund_portfolio <- fund_portfolio %>% mutate(fund_data_file_coverage = (total_mapped_value_usd)/value_usd)
+  fund_portfolio <- fund_portfolio %>% mutate(lipper_data_coverage = (total_mapped_value_usd-missing_value_usd)/total_mapped_value_usd)
+  fund_portfolio <- fund_portfolio %>% mutate(lost_coverage_fif = 1- (total_mapped_value_usd - missing_value_usd - funds_in_funds_not_mapped)/(total_mapped_value_usd-missing_value_usd))
+  
+  
+  
+  
+
+  return(fund_portfolio)
+}
+
+
+summarize_fund_coverage  <- function(fund_portfolio) {
+  
+  fund_portfolio %>% select(investor_name, 
+                            portfolio_name,
+                            isin, 
+                            value_usd,
+                            company_name,
+                            effective_coverage)
+  return(fund_portfolio)
+}
+
+list_unknown_funds_in_funds  <- function(total_portfolio) {
+  
+  fund_portfolio_total <- portfolio_total %>% filter(!is.na(fund_isin))
+  
+  table_of_funds_in_funds_not_mapped <- fund_portfolio_total %>% 
+    filter(nchar(isin)==12 & asset_type == "Funds") %>% 
+    mutate(direct_holding="FALSE")
+  
+  table_of_funds_in_funds_not_mapped <- table_of_funds_in_funds_not_mapped %>% 
+    select(investor_name, 
+           portfolio_name,
+           isin,
+           value_usd,
+           company_name,
+           fund_isin) %>% rename(parent_fund_isin = fund_isin)
```

Comment:
this whole thing could be one chain without the intermediary objects... unless those might be needed for something some day, e.g.
``` r
total_portfolio %>%
  filter(!is.na(fund_isin)) %>%
  filter(nchar(isin) == 12 & asset_type == "Funds") %>%
  mutate(direct_holding = "FALSE") %>%
  select(investor_name, 
             portfolio_name,
             isin,
             value_usd,
             company_name,
             fund_isin) %>%
  rename(parent_fund_isin = fund_isin)
```

Reply:
Done.

Reply:
this has been somewhat implemented